module/input/Camera/src/Camera.cpp

```diff
@@ -1,129 +1,364 @@
+                    if (total_local < TRANSFER_OFFSET_OBSERVE_TIME) {
+                        timesync.drift.max_clock_drift = timesync.drift.max_clock_drift == 0
+                                                             ? std::abs(total_cam - total_local) + MAX_CLOCK_DRIFT
+                                                             : (timesync.drift.max_clock_drift * 9
+                                                                + std::abs(total_cam - total_local) + MAX_CLOCK_DRIFT)
+                                                                   / 10;
+                    }
+                    else {
+                        // If our camera has drifted out by too large a margin, we recalibrate our clock
+                        timesync.drift.over_time_count +=
+                            std::abs(total_cam - total_local) > timesync.drift.max_clock_drift
+                                ? 1
+                                : (timesync.drift.over_time_count == 0 ? 0 : -1);
+
+                        // 100 frames have been over time recently
+                        if (timesync.drift.over_time_count > MAX_COUNT_OVER_TIME_FRAMES) {
+                            context->reactor.log<NUClear::INFO>(fmt::format(
+                                "Clock drift for {} camera exceeded threshold ({:.1f}ms > {:.1f}ms), recalibrating",
+                                context->name,
+                                std::abs(total_cam - total_local) / 1e6,
+                                timesync.drift.max_clock_drift / 1e6));
+
+                            arv_camera_stop_acquisition(context->camera.get());
+                            arv_stream_set_emit_signals(context->stream.get(), false);
+                            context->time = sync_clocks(arv_camera_get_device(context->camera.get()));
+                            arv_stream_set_emit_signals(context->stream.get(), true);
+                            arv_camera_start_acquisition(context->camera.get());
+                        }
+                    }
                 }
 
-                emit(msg);
-            });
+                // Apply our timestamp offset
+                ts = ts + context->time.offset;
 
-        on<Shutdown>().then([this] {
-            ShutdownV4L2Camera();
-            ShutdownAravisCamera();
-        });
-    }
+                auto msg        = std::make_unique<Image>();
+                msg->format     = context->fourcc;
+                msg->dimensions = Eigen::Matrix<unsigned int, 2, 1>(width, height);
+                // TODO(trent) use an std::vector here to avoid the copy
+                msg->data.insert(msg->data.end(), buff, buff + buffSize);
+                msg->camera_id = context->camera_id;
+                msg->name      = context->name;
+                msg->timestamp = NUClear::clock::time_point(nanoseconds(ts));
+                // TODO attach Hcw once we have access to sensors
```

Comment:
This needs to exist for this PR to be able to merge

Reply:
Working on it